File: src/gigaphone/core/plan_record.py
"""Plan record — the axis-neutral unit the fix engine consumes (DESIGN §11).

A plan record names no harness, no vendor, and no codebase specifics beyond what
discovery already wrote into config. It says only: *here is a boundary, here is what's
wrong with its span coverage, here is the complete output it should carry.* The codemod
engine routes off ``failure_modes`` to backend-adapter primitives rendered by the active
language pack.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field

from gigaphone.core.boundary import BoundaryKind, FailureMode, Source


@dataclass(frozen=True)
class PlanRecord:
    boundary: str  # "tools/exec.py:42"  (path:line)
    language: str  # "python" | "typescript" | ...
    provider_or_framework: str  # "anthropic" | "langgraph" | "acme-gateway" | ...
    kind: BoundaryKind
    tools_covered: list[str] = field(default_factory=list)
    failure_modes: list[FailureMode] = field(default_factory=list)
    complete_output_fields: list[str] = field(default_factory=list)
    source: Source = Source.ANCHOR

    def to_dict(self) -> dict:
        """JSON-ready dict; enums rendered as their string values (matches DESIGN §11)."""
        d = asdict(self)
        d["kind"] = self.kind.value
        d["failure_modes"] = [m.value for m in self.failure_modes]
        d["source"] = self.source.value
        return d

    @classmethod
    def from_dict(cls, d: dict) -> PlanRecord:
        return cls(
            boundary=d["boundary"],
            language=d["language"],
            provider_or_framework=d["provider_or_framework"],
            kind=BoundaryKind(d["kind"]),
            tools_covered=list(d.get("tools_covered", [])),
            failure_modes=[FailureMode(m) for m in d.get("failure_modes", [])],
            complete_output_fields=list(d.get("complete_output_fields", [])),
            source=Source(d.get("source", Source.ANCHOR.value)),
        )

File: src/core/planRecord.ts
/**
 * Plan record — the axis-neutral unit the fix engine consumes (DESIGN §11).
 *
 * A plan record names no harness, no vendor, and no codebase specifics beyond what discovery
 * already wrote into config. It says only: *here is a boundary, here is what's wrong with its
 * span coverage, here is the complete output it should carry.* The codemod engine routes off
 * `failureModes` to backend-adapter primitives rendered by the active language pack.
 */

import type { BoundaryKind, FailureMode, Source } from "./boundary.js";

export interface PlanRecordInit {
  boundary: string; // "tools/exec.py:42" (path:line)
  language: string; // "python" | "typescript" | ...
  providerOrFramework: string; // "anthropic" | "langgraph" | "acme-gateway" | ...
  kind: BoundaryKind;
  toolsCovered?: string[];
  failureModes?: FailureMode[];
  completeOutputFields?: string[];
  source?: Source;
}

/** JSON wire shape (snake_case keys, enums as string values) — matches DESIGN §11. */
export interface PlanRecordDict {
  boundary: string;
  language: string;
  provider_or_framework: string;
  kind: string;
  tools_covered: string[];
  failure_modes: string[];
  complete_output_fields: string[];
  source: string;
}

export class PlanRecord {
  boundary: string;
  language: string;
  providerOrFramework: string;
  kind: BoundaryKind;
  toolsCovered: string[];
  failureModes: FailureMode[];
  completeOutputFields: string[];
  source: Source;

  constructor(init: PlanRecordInit) {
    this.boundary = init.boundary;
    this.language = init.language;
    this.providerOrFramework = init.providerOrFramework;
    this.kind = init.kind;
    this.toolsCovered = init.toolsCovered ?? [];
    this.failureModes = init.failureModes ?? [];
    this.completeOutputFields = init.completeOutputFields ?? [];
    this.source = init.source ?? "anchor";
  }

  toDict(): PlanRecordDict {
    return {
      boundary: this.boundary,
      language: this.language,
      provider_or_framework: this.providerOrFramework,
      kind: this.kind,
      tools_covered: [...this.toolsCovered],
      failure_modes: [...this.failureModes],
      complete_output_fields: [...this.completeOutputFields],
      source: this.source,
    };
  }

  static fromDict(d: PlanRecordDict): PlanRecord {
    return new PlanRecord({
      boundary: d.boundary,
      language: d.language,
      providerOrFramework: d.provider_or_framework,
      kind: d.kind as BoundaryKind,
      toolsCovered: [...(d.tools_covered ?? [])],
      failureModes: [...((d.failure_modes ?? []) as FailureMode[])],
      completeOutputFields: [...(d.complete_output_fields ?? [])],
      source: (d.source ?? "anchor") as Source,
    });
  }
}
